dist/gaia-header.js

```diff
@@ -277,124 +456,519 @@ module.exports = Component.register('gaia-header', {
+
+    var titles = this.els.titles;
+    var space = this.getTitleSpace();
+    var styles = [].map.call(titles, (el) => this.getTitleStyle(el, space));
+
+    // Update the title styles using the latest
+    // styles. This function can be called many
+    // times but will only run once in each tick.
+    return this.setTitleStylesSoon(styles);
+  },
+
+  /**
+   * Microtask debounced `runFontFit`
    *
-   * TODO: We really need an official API for this.
+   * @private
+   */
+  runFontFitSoon: function() {
+    debug('run font-fit soon');
+    if (this.pending.runFontFitSoon) { return; }
+    this.pending.runFontFitSoon = this.nextTick(() => {
+      delete this.pending.runFontFitSoon;
+      this.runFontFit();
+    });
+  },
+
+  /**
+   * Get the style properties required
+   * to fit and position the title text.
    *
+   * @param  {HTMLH1Element} el
+   * @param  {Number} space
+   * @return {Object} {fontSize, marginStart, overflowing, id}
    * @private
    */
-  rerunFontFit: function() {
-    for (var i = 0; i < this.els.headings.length; i++) {
-      fontFit.reformatHeading(this.els.headings[i]);
+  getTitleStyle: function(el, space) {
+    debug('get el style', el, space);
+    var text = el.textContent;
+    var styleId = text + space;
+
+    // Bail when there's no text (or just whitespace)
+    if (!text || !text.trim()) { return false; }
+
+    // If neither the text or the titleSpace
+    // changed, there's no reason to continue.
+    if (getStyleId(el) === styleId) { return false; }
+
+    var marginStart = this.getTitleMarginStart();
+    var textSpace = space - Math.abs(marginStart);
+    var fontFitResult = this.fontFit(text, textSpace);
+    var overflowing = fontFitResult.textWidth > textSpace;
+
+    // If the text is overflowing in the
+    // centered title, we remove centering
+    // to free up space, rerun fontFit to
+    // get a fontSize which fits this space.
+    if (overflowing) {
+      fontFitResult = this.fontFit(text, space);
+      marginStart = 0;
     }
+
+    return {
+      fontSize: fontFitResult.fontSize,
+      marginStart: marginStart,
+      overflowing: overflowing,
+      id: styleId
+    };
   },
 
   /**
-   * Triggers the 'action' button
-   * (used in testing).
+   * Set's styles on the title elements.
    *
-   * @public
+   * If there is already an unresolved Promise
+   * we return it instead of scheduling another.
+   * This means that the function can be hammered
+   * in the same sync-turn and will only run
+   * once (like a debounce).
+   *
+   * @param {Array} styles
+   * @return {Promise}
    */
-  triggerAction: function() {
-    if (this.isSupportedAction(this.getAttribute('action'))) {
-      this.els.actionButton.click();
-    }
+  setTitleStylesSoon: function(styles) {
+    debug('set title styles soon', styles);
+    var key = 'setStyleTitlesSoon';
+
+     // Always update styles
+    this._titleStyles = styles;
```

Comment:
Maybe I'm confused, here is what I have in mind:
```javascript
function defer() {
  var deferred = {};
  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve;
    deferred.reject = reject;
  });

  return deferred;
}

setTitleStylesSoon: function(styles) {
  if (this.pending[key]) {
    this.pending[key].clear();
  }

  this.unresolved[key] = this.unresolved[key] || defer();
  
  this.pending[key] = this.nextTick(() => {
    var els = this.els.titles;

    [].forEach.call(els, (el, i) => {
      if (!styles[i]) { return debug('exit'); }
      this.setTitleStyle(el, styles[i]);
    });

    this.unresolved[key].resolve();

    // Clean up
    delete this._titleStyles;
    delete this.unresolved[key];
    delete this.pending[key];
  });

  return this.unresolved[key].promise;
}
```

Basically the same function except we don't need the `this._titleStyles` var because instead of keeping the first call we keep the last call (by clearing the older pending one and still keeping the oldest promise by using a defer -- which makes it easier to manage the "resolve" function).

I think this brings the same idea but maybe I'm confused, tell me what you think.